File: Keppnisforritun/vika4/ingestion.cpp
#include <iostream>

int geymsla[200][50000];
int meals[200];

int food(int i, int r, int last, int orgKal, int orgFjoldi){
    //std::cout << i << std::endl;
    //std::cout << r << std::endl;
    //std::cout << orgKal << std::endl;
    //std::cout << orgFjoldi << std::endl;
    if (i >= orgFjoldi){
        return 0;
    }
    if(geymsla[i][r] == 0){
        geymsla[i][r] = std::max((food(i+1, r*2/3, r, orgKal, orgFjoldi) + std::min(meals[i], r)), 
                std::max(food(i+1, last, r, orgKal, orgFjoldi), food(i+2, orgKal, r, orgKal, orgFjoldi)));     
    }
    return geymsla[i][r];
}

int main()
{
	int fjoldi;
    int kaloriur;
	std::cin >> fjoldi >> kaloriur;
    for(int i = 0; i < fjoldi; i++){
        std::cin >> meals[i];
    }
    int max = food(0, kaloriur, kaloriur, kaloriur, fjoldi);
    std::cout << max << std::endl;
	return 0;
}

File: Keppnisforritun/vika4/ingestion.py
from fractions import Fraction
import math
from time import time

rows, cols = (200, 60000)
geymsla = [[0]*cols]*rows

data = input().split()
n = int(data[0])
m = int(data[1])
data = input().split()
meals = [int(i) for i in data]


#print("meals:", meals)


#i=index, r=ratio af m, last=síðasta ratio
def food(r, i, last):
    #print("index:",i,"ratio:", r, "ratio*(2/3):", int(r*(2/3)), "last:", last)
    if i >= n:
        return 0
    if geymsla[i][r] == 0:
        # 1= Eat, 2= NoEat, 3= twoNoEat
        theFood = int(min(meals[i], r))
        geymsla[i][r] = max((food(int(r*2/3),i + 1, r) + min(meals[i], r)), 
                max(food(last, i + 1, r), food( m, i + 2, r)))
    #print("nae i ur geymslu, [i][r]:", geymsla[i][r])
    return geymsla[i][r]
    

max = food(m, 0, m)

print(max)

# 3 900
# 400 1000 1500
# kemur rangt út














# ratio = [1]; temp = 1
# for x in range(lenMeals):
#     temp = temp * (2/3)
#     ratio.append(temp)

# def food(i, timeSince, r, sum):
#     if i == lenMeals:
#         return sum
#     else:
#         if timeSince >= 1:
#             noEat = food(i+1, timeSince+1, 0, sum)
#         else:
#             noEat = food(i+1, timeSince+1, r-1, sum)
#         if m*ratio[r] >= meals[i]:
#             eat = food(i+1, 0, r+1, sum + math.floor(meals[i]))
#         else:
#             eat = food(i+1, 0, r+1, sum + math.floor(m*ratio[r]))
#         return max(eat, noEat)

# max = food(0,2,0,0)
# print(max)

